exchange/bitswap/wantmanager.go

```diff
@@ -0,0 +1,262 @@
+		peers:      make(map[peer.ID]*msgQueue),
+		wl:         wantlist.New(),
+		network:    network,
+		ctx:        ctx,
+	}
+}
+
+type msgPair struct {
+	to  peer.ID
+	msg bsmsg.BitSwapMessage
+}
+
+type cancellation struct {
+	who peer.ID
+	blk u.Key
+}
+
+type msgQueue struct {
+	p peer.ID
+
+	outlk sync.Mutex
+	out   bsmsg.BitSwapMessage
+
+	work chan struct{}
+	done chan struct{}
+}
+
+func (pm *WantManager) WantBlocks(ks []u.Key) {
+	log.Infof("want blocks: %s", ks)
+	pm.addEntries(ks, false)
+}
+
+func (pm *WantManager) CancelWants(ks []u.Key) {
+	pm.addEntries(ks, true)
+}
+
+func (pm *WantManager) addEntries(ks []u.Key, cancel bool) {
+	var entries []*bsmsg.Entry
+	for i, k := range ks {
+		entries = append(entries, &bsmsg.Entry{
+			Cancel: cancel,
+			Entry: wantlist.Entry{
+				Key:      k,
+				Priority: kMaxPriority - i,
+			},
+		})
+	}
+	select {
+	case pm.incoming <- entries:
+	case <-pm.ctx.Done():
+	}
+}
+
+func (pm *WantManager) SendBlock(ctx context.Context, env *engine.Envelope) {
+	// Blocks need to be sent synchronously to maintain proper backpressure
+	// throughout the network stack
+	defer env.Sent()
+
+	msg := bsmsg.New(false)
+	msg.AddBlock(env.Block)
+	log.Infof("Sending block %s to %s", env.Peer, env.Block)
+	err := pm.network.SendMessage(ctx, env.Peer, msg)
+	if err != nil {
+		log.Error(err)
+	}
+}
+
+func (pm *WantManager) startPeerHandler(p peer.ID) *msgQueue {
+	_, ok := pm.peers[p]
+	if ok {
+		// TODO: log an error?
+		return nil
+	}
+
+	mq := newMsgQueue(p)
+
+	// new peer, we will want to give them our full wantlist
+	fullwantlist := bsmsg.New(true)
+	for _, e := range pm.wl.Entries() {
+		fullwantlist.AddEntry(e.Key, e.Priority)
+	}
+	mq.out = fullwantlist
+	mq.work <- struct{}{}
+
+	pm.peers[p] = mq
+	go pm.runQueue(mq)
+	return mq
+}
+
+func (pm *WantManager) stopPeerHandler(p peer.ID) {
+	pq, ok := pm.peers[p]
+	if !ok {
+		// TODO: log error?
+		return
+	}
+
+	close(pq.done)
+	delete(pm.peers, p)
+}
+
+func (pm *WantManager) runQueue(mq *msgQueue) {
+	for {
+		select {
+		case <-mq.work: // there is work to be done
+
+			err := pm.network.ConnectTo(pm.ctx, mq.p)
+			if err != nil {
+				log.Errorf("cant connect to peer %s: %s", mq.p, err)
+				// TODO: cant connect, what now?
+				continue
+			}
+
+			// grab outgoing message
+			mq.outlk.Lock()
+			wlm := mq.out
+			if wlm == nil || wlm.Empty() {
+				mq.outlk.Unlock()
+				continue
+			}
+			mq.out = nil
+			mq.outlk.Unlock()
+
+			// send wantlist updates
+			err = pm.network.SendMessage(pm.ctx, mq.p, wlm)
+			if err != nil {
+				log.Error("bitswap send error: ", err)
+				// TODO: what do we do if this fails?
+			}
+		case <-mq.done:
+			return
+		}
+	}
+}
+
+func (pm *WantManager) Connected(p peer.ID) {
+	pm.connect <- p
+}
+
+func (pm *WantManager) Disconnected(p peer.ID) {
+	pm.disconnect <- p
+}
+
+// TODO: use goprocess here once i trust it
+func (pm *WantManager) Run() {
+	tock := time.NewTicker(rebroadcastDelay.Get())
```

Comment:
AFAIK, `Tick` is bad if used in anything but a static (once-per-process) context.